File: protocol-generator/csharp/test/EntityCore.Protocol.Conformance/Corpus.cs
namespace EntityCore.Protocol.Conformance;

/// <summary>Locates the vendored conformance fixture relative to the build output.</summary>
public static class Corpus
{
    private const string RelativePath =
        "protocol-generator/shared/test-vectors/v0.8.0/conformance-vectors-v1.cbor";

    /// <summary>
    /// Walk up from the current directory to find the repo's vendored fixture.
    /// Falls back to the <c>ECF_VECTORS</c> environment variable if set.
    /// </summary>
    public static string Locate()
    {
        string? fromEnv = Environment.GetEnvironmentVariable("ECF_VECTORS");
        if (!string.IsNullOrEmpty(fromEnv) && File.Exists(fromEnv))
        {
            return fromEnv;
        }

        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            string candidate = Path.Combine(dir.FullName, RelativePath);
            if (File.Exists(candidate))
            {
                return candidate;
            }
            dir = dir.Parent;
        }

        throw new FileNotFoundException(
            $"could not locate {RelativePath} above {AppContext.BaseDirectory}; " +
            "set ECF_VECTORS to the fixture path.");
    }
}

File: protocol-generator/typescript/test/corpus.ts
import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

/**
 * Locates the vendored conformance fixture relative to the build output (the
 * keystone repo layout). Twin of the C# `Corpus.Locate`. `ECF_VECTORS` overrides.
 */
const RELATIVE_PATH = "protocol-generator/shared/test-vectors/v0.8.0/conformance-vectors-v1.cbor";

export function locateCorpus(): string {
  const fromEnv = process.env["ECF_VECTORS"];
  if (fromEnv && existsSync(fromEnv)) {
    return fromEnv;
  }
  let dir = dirname(fileURLToPath(import.meta.url));
  // Walk up to the repo root looking for the vendored fixture.
  for (let i = 0; i < 12; i++) {
    const candidate = join(dir, RELATIVE_PATH);
    if (existsSync(candidate)) {
      return candidate;
    }
    const parent = dirname(dir);
    if (parent === dir) {
      break;
    }
    dir = parent;
  }
  throw new Error(`could not locate ${RELATIVE_PATH} above ${fileURLToPath(import.meta.url)}; set ECF_VECTORS`);
}

export function loadCorpusBytes(): Uint8Array {
  return new Uint8Array(readFileSync(locateCorpus()));
}
